File: AdventureWorks Data Binding sample/[C#]-AdventureWorks Data Binding sample/C#/AdWksSalesWinDataBind/SalesOrderHelpers.cs
///////////////////////////////////////////////////////////////
//
// This code is part of the AdventureWorks Data Binding sample.
//
// Copyright (c) Microsoft Corporation.  All rights reserved.
//
///////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Data;
using System.Data.Objects;
using AdWksSalesWinDataBind.Properties;

namespace AdWksSalesWinDataBind
{
    public partial class SalesOrderHeader
    {
        // Update the order total.
        public void UpdateOrderTotal()
        {
            decimal newSubTotal = 0;

            // In a real world application this information should be included in the EDM.
            decimal taxRate = Settings.Default.currentSalesTaxRate;
            decimal freight = Settings.Default.currentFreight;

            // If the items for this order are loaded or if the order is 
            // newly added, then recalculate the subtotal as it may have changed.
            if (SalesOrderDetail.IsLoaded ||
                EntityState == EntityState.Added)
            {
                foreach (SalesOrderDetail item in this.SalesOrderDetail)
                {
                    // Calculate line totals for loaded items.
                    newSubTotal += item.LineTotal;
                }

                this.SubTotal = newSubTotal;
            }

            // Calculate the new tax amount.
            this.TaxAmt = this.SubTotal * taxRate;

            // Calculate the new freight amount.
            this.Freight = this.SubTotal * freight;

            // Calculate the new total.
            this.TotalDue = this.SubTotal + this.TaxAmt + this.Freight;
        }
    }
    public partial class SalesOrderDetail
    {
        // Update the line total.
        public void UpdateLineTotal()
        {
            this.LineTotal = this.OrderQty *
                (this.UnitPrice * (1 - this.UnitPriceDiscount));

            if (this.SalesOrderHeader != null)
            {
                // Ensure that we update the totals in the related order.
                this.SalesOrderHeader.UpdateOrderTotal();
            }
        }
        partial void OnOrderQtyChanging(short value)
        {
            // Validate the supplied value.
            if (value < 1)
            {
                throw new ApplicationException("An order must be 1 or more items.");
            }
        }
        partial void OnOrderQtyChanged()
        {
            // Update the line total to reflect the new quantity;
            this.UpdateLineTotal();
        }
        partial void OnUnitPriceDiscountChanging(decimal value)
        {
            // Validate the supplied value.
            if (value < 0 || value > 1)
            {
                throw new ApplicationException("Discount rate is between 0 and 1.");
            }
        }
        partial void OnUnitPriceDiscountChanged()
        {
            // Update the line total to reflect the new discount;
            this.UpdateLineTotal();
        }
    }
}

File: AdventureWorks Data Binding sample/[VB]-AdventureWorks Data Binding sample/VB/AdWksSalesWinDataBindVb/SalesOrderHelpers.vb

Partial Public Class SalesOrderHeader
    Public Sub UpdateOrderTotal()
        Dim newSubTotal As Decimal

        'In a real world application this information should be included in the EDM.
        Dim taxRate As Decimal = My.Settings.currentSalesTaxRate
        Dim freight As Decimal = My.Settings.currentFreight

        'If the items for this order are loaded or if the order is
        'newly added, then recalculate the subtotal as it may have changed.
        If SalesOrderDetail.IsLoaded OrElse EntityState = Data.EntityState.Added Then
            For Each item In SalesOrderDetail
                'Calculate line totals for loaded items
                newSubTotal += item.LineTotal
            Next
            SubTotal = newSubTotal
        End If

        'Calculate the new tax amount
        TaxAmt = SubTotal * taxRate

        'calculate the new total
        TotalDue = SubTotal + TaxAmt + freight
    End Sub
End Class

Partial Public Class SalesOrderDetail
    Public Sub UpdateLineTotal()
        LineTotal = OrderQty * (UnitPrice * (1 - UnitPriceDiscount))

        If SalesOrderHeader IsNot Nothing Then
            'Ensure that we update the totals in the related order.
            SalesOrderHeader.UpdateOrderTotal()
        End If
    End Sub

    Private Sub OnOrderQtyChanging(ByVal value As Short)
        'validate the supplied value
        If value < 1 Then
            Throw New ApplicationException("An order must be 1 or more items.")
        End If
    End Sub

    Private Sub OnOrderQtyChanged()
        'Update the line total to reflect the new quantity
        UpdateLineTotal()
    End Sub

    Private Sub OnUnitPriceDiscountChanging(ByVal value As Decimal)
        'Validate the supplied value
        If value < 0 OrElse value > 1 Then
            Throw New ApplicationException("Discount rate is between 0 and 1.")
        End If
    End Sub

    Private Sub OnUnitPriceDiscountChanged()
        'Update the line total to reflect the new discount
        UpdateLineTotal()
    End Sub
End Class
